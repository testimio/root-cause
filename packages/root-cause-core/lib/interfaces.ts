import type { Page as PuppeteerPage } from 'puppeteer';
import type { Page as PlaywrightPage } from 'playwright';
import type { TestContextInterface } from './TestContext';
import type { TestEndStatus } from './attachInterfaces';
import type { InstrumentedFunctionResult } from '@testim/root-cause-types';

export type RootCausePage = PuppeteerPage | PlaywrightPage;

export interface AttachOptions {
  noResultsServer?: boolean;
  externalTestName?: string;
  resultsDirectory?: string;
}

export interface CommonEachHookArgs {
  testContext: TestContextInterface;
  fnName: string;
  proxyContext: any;
  rootPage: RootCausePage;
  args: any[];
  methodCallData: ProxiedMethodCallData[];
}

export type BeforeHookArgs = CommonEachHookArgs;

export type BeforeHook = (args: BeforeHookArgs) => Promise<void>;

export interface AfterHookArgs extends CommonEachHookArgs {
  instrumentedFunctionResult: InstrumentedFunctionResult<any, any>;
}

export type AfterHook = (args: AfterHookArgs) => Promise<void>;

export interface BeforeAllHookArgs {
  testContext: TestContextInterface;
  proxyContext: any;
  rootPage: RootCausePage;
}

export type BeforeAllHook = (args: BeforeAllHookArgs) => Promise<void>;

export interface AfterAllHookArgs {
  testContext: TestContextInterface;
  endStatus: TestEndStatus<unknown, unknown>;
}

export type AfterAllHook = (args: AfterAllHookArgs) => Promise<void>;

export interface IAutomationFrameworkInstrumentor {
  // register additional hooks on a context, this is implemented for
  // puppeteer/selenium
  // registerHooks(page: T, hooks: AutomationFrameworkHooks<T>): T;
  registerBeforeAllHook(hook: BeforeAllHook): void;
  registerAfterAllHook(hook: AfterAllHook): void;
  registerBeforeHook(hook: BeforeHook): void;
  registerAfterHook(hook: AfterHook): void;
  wrapWithProxy<T extends object>(proxiedObject: T): T;
  end(endStatus: TestEndStatus<unknown, unknown>): Promise<void>;
  pause(): void;
  resume(): void;
}

export interface ProxiedMethodCallData {
  creationFunction: string;
  selector?: string;
  text?: string;
  index?: number;
}
