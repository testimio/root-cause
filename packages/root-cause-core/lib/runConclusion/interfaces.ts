import { TestResultFile } from '@testim/root-cause-types';

export interface RootCauseRunResultEntry {
  singleResultDir: string;
  id: string;
  testData: TestResultFile;
}
