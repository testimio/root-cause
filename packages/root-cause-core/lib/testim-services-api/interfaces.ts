// generated with json2ts

interface TestimExecutionExecutionRunConfig {
  parallel: number;
  browser: string;
  runnerVersion: string;
  testimBranch: string;
  canaryMode: boolean;
  source: string;
  testPlans: any[];
  testLabels: any[];
  testNames: any[];
  testIds: any[];
  testConfigs: any[];
  testConfigIds: any[];
  browserTimeout: number;
  timeout: number;
  newBrowserWaitTimeout: number;
  tunnelPort: string;
  runnerMode: string;
  sessionType: string;
  companyId: string;
  testData?: any;
  testDataTotal?: any;
  testDataIndex?: any;
  testConfig: {}; // not too interesting
}

interface TestExecution {
  testId: string;
  status: string;
  name: string;
  resultId: string;
  isTestsContainer: boolean;
  testRunOverrideStatus: string;
  config: TestimExecutionExecutionRunConfig;
  success?: boolean;
  reason?: string;
  childTestResultIds?: string[];
  startTime: number;
  endTime: number;
  parentResultId?: string;
  show?: boolean;
}

export interface TestimBackendExecutionFormatSubsetForReporting {
  runId: string;
  execution: {
    [id: string]: {
      testId: string;
      name: string;
      success?: boolean;
      resultId: string;
      parentResultId?: string;
      isTestsContainer: boolean;
    };
  };
  startTime: number;
}
export interface TestimBackendExecutionInputFormat {
  runId: string;
  projectId: string;
  labels: string;
  startTime: number;
  endTime: number;
  execution: { [id: string]: TestExecution };
  status: string;
  config: TestimExecutionExecutionRunConfig;
  resultLabels: string[];
  remoteRunId: string;
  metadata: any;
  show: boolean;
}
