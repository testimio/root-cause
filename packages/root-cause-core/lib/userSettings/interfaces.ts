export interface PossibleUserSettings {
  features?: Partial<FeaturesSettings>;
}

export interface FeaturesSettings {
  /**
   * Take screenshot for every step
   *
   * Default to true, with jpeg 85
   */
  screenshots?: boolean | ScreenshotsSettings;

  /**
   * Record the browser's console
   *
   * @default true
   */
  console?: boolean;

  /**
   * Record network activity, make it available in the UI and as har file
   */
  networkLogs?: boolean;

  /**
   * When using jest integration, record jest expect calls as assertion steps
   */
  jestAssertions?: boolean;

  /**
   * Record the DOM content of the page for every step
   */
  html?: boolean;
}

export interface ResolvedSettings {
  features: {
    screenshots: false | Required<ScreenshotsSettings>;
    console: boolean;
    networkLogs: boolean;
    jestAssertions: boolean;
    html: boolean;
  };
}

export interface ScreenshotsSettings {
  format: 'png' | 'jpeg';

  /**
   * Quality of jpeg image, from 1 to 100
   *
   * @default 85
   */
  quality?: number;

  /**
   * When true, takes a screenshot of the full scrollable page, instead of the currently visible
   *
   * @default false
   */
  fullPage?: boolean;
}

type ContentType = string;

/**
 * placeholder, not in use yet
 */
export interface NetworkLogsSettings {
  /**
   * Save the response bodies into the har file, or only for specific content types
   *
   * @default false
   */
  includeResponseBody?: boolean | ContentType[];
}
