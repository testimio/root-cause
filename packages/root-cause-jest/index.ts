export * from './lib/helpers';
export * from './lib/jasmine2JestRelated';
