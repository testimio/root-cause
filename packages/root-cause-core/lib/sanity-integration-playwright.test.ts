import {
  getCleanAllPathsPrettyFormatPlugin,
  getCleanProcessTicksAndRejectionsStackFramePrettyFormatPlugin,
} from '@testim/internal-self-tests-helpers';
// eslint-disable-next-line import/no-extraneous-dependencies
import { getJasmineCurrentTest, registerJasmineReporterToGlobal } from '@testim/root-cause-jest';
import type { TestResultFile } from '@testim/root-cause-types';
import assert from 'assert';
import fs from 'fs-extra';
import path from 'path';
import playwright from 'playwright';
import { attach } from './index';
import { PossibleUserSettings } from './userSettings/interfaces';
import { resolveSettings } from './userSettings/userSettings';
import {
  assertNotNullOrUndefined,
  jsonReduceNoiseReviver,
  testUniqueIdentifierFromStartParams,
} from './utils';

describe('Sanity integration test playwright', () => {
  jest.setTimeout(30_000);
  expect.addSnapshotSerializer(getCleanAllPathsPrettyFormatPlugin(process.cwd()));
  expect.addSnapshotSerializer(getCleanProcessTicksAndRejectionsStackFramePrettyFormatPlugin());

  let browser: playwright.Browser;
  let browserContext: playwright.BrowserContext;
  let page: playwright.Page;

  beforeAll(async () => {
    registerJasmineReporterToGlobal();

    // https://github.com/puppeteer/puppeteer/blob/master/docs/troubleshooting.md#setting-up-chrome-linux-sandbox
    // Not optimal, but didn't work on circle on circleci/node:12.17-stretch-browsers without it
    // Need to revisit
    browser = await playwright.chromium.launch({
      args: ['--no-sandbox', '--disable-setuid-sandbox'],
    });
  });

  afterAll(async () => {
    await browser.close();
  });

  beforeEach(async () => {
    browserContext = await browser.newContext();
    page = await browserContext.newPage();
  });

  afterEach(async () => {
    await page.close();
    await browserContext.close();
  });

  it('Sanity integration test 1', async () => {
    const currentTestInfo = getJasmineCurrentTest();

    const startTestParams = {
      runId: 'mock_invocation_id-playwright',
      projectRoot: path.resolve(__dirname, 'testsResults'),
      fullName: currentTestInfo.fullName,
      description: currentTestInfo.description,
      fullSuitePath: __filename,
    };

    const mockedDateConstructor: typeof Date = Object.create(Date);

    let nowCallsCounter = 1;
    mockedDateConstructor.now = function mockedNow() {
      return nowCallsCounter++;
    };

    const overriddenFeatures: PossibleUserSettings = {
      features: {
        html: true,
      },
    };

    const { page: playedPage, endTest } = await attach(
      {
        page,
        startTestParams,
        activeFeatures: resolveSettings(overriddenFeatures).features,
      },
      mockedDateConstructor
    );

    await playedPage.goto('http://jsbin.testim.io/ner/1');
    await playedPage.click('#test');
    // to await - might be flaky
    page.evaluate(() => {
      setTimeout(() => {
        throw new Error('injected error');
      }, 0);
    });
    await playedPage.click('#test');
    await playedPage.click('#test');
    // to await - might be flaky
    page.evaluate(() => {
      setTimeout(() => {
        /* eslint-disable-next-line no-console */
        console.warn('injected console');
      }, 0);
    });
    await playedPage.click('#test');
    const counterEl = await playedPage.$('#counter');

    assert.notEqual(counterEl, null);
    assertNotNullOrUndefined(counterEl);

    const text = await (await counterEl.getProperty('innerText')).jsonValue();

    assert.equal(text, '4');

    await playedPage.click('#reset');

    const text2 = await (await counterEl.getProperty('innerText')).jsonValue();

    assert.equal(text2, '0');

    const allButtons = await playedPage.$$('button');

    await allButtons[0].click();

    const text3 = await (await counterEl.getProperty('innerText')).jsonValue();

    assert.equal(text3, '1');

    assert.equal(allButtons.length, 2);

    await endTest({
      success: false,
      error: { message: 'mocked error test failed' },
    });

    const expectedResults = path.resolve(
      __dirname,
      'testsResults',
      '.root-cause',
      'runs',
      startTestParams.runId,
      testUniqueIdentifierFromStartParams(startTestParams)
    );

    const testResults: TestResultFile = JSON.parse(
      await fs.readFile(path.resolve(expectedResults, 'results.json'), 'utf-8'),
      jsonReduceNoiseReviver
    );

    expect(testResults).toMatchSnapshot();

    for (const stepResult of testResults.steps) {
      // Do also images blob comparison? TBH
      if (stepResult.screenshot) {
        assert.equal(
          await fs.pathExists(path.resolve(expectedResults, stepResult.screenshot)),
          true,
          `${stepResult.screenshot} is missing`
        );
      }
    }
  });
});
