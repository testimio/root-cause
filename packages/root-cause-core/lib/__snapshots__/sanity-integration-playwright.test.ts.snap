// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sanity integration test playwright Sanity integration test 1 1`] = `
Object {
  "metadata": Object {
    "branchInfo": "noise_removed:branchInfo",
    "endedTimestamp": 29,
    "fileName": "noise_removed/packages/root-cause-core/lib/sanity-integration-playwright.test.ts",
    "hasNetworkLogs": true,
    "systemInfo": Object {
      "automationFramework": "playwright",
      "browser": "chromium",
      "browserPlatform": "noise_removed:browserPlatform",
      "browserVersion": "N/A",
      "modelName": "noise_removed:modelName",
      "modelVersion": "noise_removed:modelVersion",
      "pageViewport": Object {
        "height": 720,
        "width": 1280,
      },
      "userAgent": "noise_removed:userAgent",
    },
    "testEndStatus": Object {
      "codeLocationDetails": Object {
        "callstack": Array [
          Object {
            "column": 29,
            "file": "lib/assortedHooks.ts",
            "function": "testEndHook",
            "line": 116,
          },
          Object {
            "column": 15,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "PuppeteerPageHooker.end",
            "line": 229,
          },
          Object {
            "column": 5,
            "file": "lib/attach.ts",
            "function": "endTest",
            "line": 115,
          },
          Object {
            "column": 5,
            "file": "lib/sanity-integration-playwright.test.ts",
            "function": "Object.<anonymous>",
            "line": 131,
          },
        ],
        "codeLines": Array [
          "",
          "    assert.equal(allButtons.length, 2);",
          "",
          "    await endTest({",
          "      success: false,",
          "      error: { message: 'mocked error test failed' },",
          "    });",
        ],
        "column": 5,
        "fromRowNumber": 128,
        "row": 131,
        "sourceFileRelativePath": "lib/sanity-integration-playwright.test.ts",
        "toRowNumber": 134,
      },
      "error": Object {
        "message": "mocked error test failed",
        "name": "",
        "stack": "",
      },
      "success": false,
    },
    "testFullName": "Sanity integration test playwright Sanity integration test 1",
    "testName": "Sanity integration test 1",
    "timestamp": 2,
  },
  "steps": Array [
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 4,
      "fnName": "goto",
      "index": 1,
      "mhtmlFile": "1.document.mhtml",
      "name": "goto \\"http://jsbin.testim.io/ner/1\\"",
      "pageId": 0,
      "screenshot": "1.page-goto.jpg",
      "startTimestamp": 3,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_Page_goto",
            "line": 196,
          },
          Object {
            "column": 5,
            "file": "lib/sanity-integration-playwright.test.ts",
            "function": "Object.<anonymous>",
            "line": 88,
          },
        ],
        "codeLines": Array [
          "      mockedDateConstructor",
          "    );",
          "",
          "    await playedPage.goto('http://jsbin.testim.io/ner/1');",
          "    await playedPage.click('#test');",
          "    // to await - might be flaky",
          "    page.evaluate(() => {",
        ],
        "column": 5,
        "fromRowNumber": 85,
        "row": 88,
        "sourceFileRelativePath": "lib/sanity-integration-playwright.test.ts",
        "toRowNumber": 91,
      },
      "text": "http://jsbin.testim.io/ner/1",
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 6,
      "fnName": "click",
      "index": 2,
      "mhtmlFile": "2.document.mhtml",
      "name": "click \\"#test\\"",
      "pageId": 0,
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "2.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 5,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_Page_click",
            "line": 196,
          },
          Object {
            "column": 5,
            "file": "lib/sanity-integration-playwright.test.ts",
            "function": "Object.<anonymous>",
            "line": 89,
          },
        ],
        "codeLines": Array [
          "    );",
          "",
          "    await playedPage.goto('http://jsbin.testim.io/ner/1');",
          "    await playedPage.click('#test');",
          "    // to await - might be flaky",
          "    page.evaluate(() => {",
          "      setTimeout(() => {",
        ],
        "column": 5,
        "fromRowNumber": 86,
        "row": 89,
        "sourceFileRelativePath": "lib/sanity-integration-playwright.test.ts",
        "toRowNumber": 92,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 9,
      "fnName": "click",
      "index": 3,
      "mhtmlFile": "3.document.mhtml",
      "name": "click \\"#test\\"",
      "pageId": 0,
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "3.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 7,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_Page_click",
            "line": 196,
          },
          Object {
            "column": 5,
            "file": "lib/sanity-integration-playwright.test.ts",
            "function": "Object.<anonymous>",
            "line": 96,
          },
        ],
        "codeLines": Array [
          "        throw new Error('injected error');",
          "      }, 0);",
          "    });",
          "    await playedPage.click('#test');",
          "    await playedPage.click('#test');",
          "    // to await - might be flaky",
          "    page.evaluate(() => {",
        ],
        "column": 5,
        "fromRowNumber": 93,
        "row": 96,
        "sourceFileRelativePath": "lib/sanity-integration-playwright.test.ts",
        "toRowNumber": 99,
      },
      "unhandledExceptions": Array [
        Object {
          "message": "injected error",
          "stack": "Error: injected error
    at eval (eval at callFunction (:307:29), <anonymous>:3:23)",
          "timestamp": 8,
        },
      ],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 11,
      "fnName": "click",
      "index": 4,
      "mhtmlFile": "4.document.mhtml",
      "name": "click \\"#test\\"",
      "pageId": 0,
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "4.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 10,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_Page_click",
            "line": 196,
          },
          Object {
            "column": 5,
            "file": "lib/sanity-integration-playwright.test.ts",
            "function": "Object.<anonymous>",
            "line": 97,
          },
        ],
        "codeLines": Array [
          "      }, 0);",
          "    });",
          "    await playedPage.click('#test');",
          "    await playedPage.click('#test');",
          "    // to await - might be flaky",
          "    page.evaluate(() => {",
          "      setTimeout(() => {",
        ],
        "column": 5,
        "fromRowNumber": 94,
        "row": 97,
        "sourceFileRelativePath": "lib/sanity-integration-playwright.test.ts",
        "toRowNumber": 100,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [
        Object {
          "args": Array [
            "\\"injected console\\"",
          ],
          "column": 24,
          "level": "warning",
          "line": 3,
          "text": "injected console",
          "timestamp": 13,
          "url": "",
        },
      ],
      "endTimestamp": 14,
      "fnName": "click",
      "index": 5,
      "mhtmlFile": "5.document.mhtml",
      "name": "click \\"#test\\"",
      "pageId": 0,
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "5.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 12,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_Page_click",
            "line": 196,
          },
          Object {
            "column": 5,
            "file": "lib/sanity-integration-playwright.test.ts",
            "function": "Object.<anonymous>",
            "line": 105,
          },
        ],
        "codeLines": Array [
          "        console.warn('injected console');",
          "      }, 0);",
          "    });",
          "    await playedPage.click('#test');",
          "    const counterEl = await playedPage.$('#counter');",
          "",
          "    assert.notEqual(counterEl, null);",
        ],
        "column": 5,
        "fromRowNumber": 102,
        "row": 105,
        "sourceFileRelativePath": "lib/sanity-integration-playwright.test.ts",
        "toRowNumber": 108,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 16,
      "fnName": "$",
      "index": 6,
      "mhtmlFile": "6.document.mhtml",
      "name": "$ \\"#counter\\"",
      "pageId": 0,
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "6.page-$.jpg",
      "selector": "#counter",
      "startTimestamp": 15,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 30,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "Proxy.returningElementHandlesViaPromiseWrappedFunction_Page_$",
            "line": 159,
          },
          Object {
            "column": 23,
            "file": "lib/sanity-integration-playwright.test.ts",
            "function": "Object.<anonymous>",
            "line": 106,
          },
        ],
        "codeLines": Array [
          "      }, 0);",
          "    });",
          "    await playedPage.click('#test');",
          "    const counterEl = await playedPage.$('#counter');",
          "",
          "    assert.notEqual(counterEl, null);",
          "    assertNotNullOrUndefined(counterEl);",
        ],
        "column": 23,
        "fromRowNumber": 103,
        "row": 106,
        "sourceFileRelativePath": "lib/sanity-integration-playwright.test.ts",
        "toRowNumber": 109,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 18,
      "fnName": "getProperty",
      "index": 7,
      "mhtmlFile": "7.document.mhtml",
      "name": "getProperty \\"innerText\\" on \\"#counter\\"",
      "pageId": 0,
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "7.elementhandle-getProperty.jpg",
      "selector": "#counter",
      "startTimestamp": 17,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_ElementHandle_getProperty",
            "line": 196,
          },
          Object {
            "column": 25,
            "file": "lib/sanity-integration-playwright.test.ts",
            "function": "Object.<anonymous>",
            "line": 111,
          },
        ],
        "codeLines": Array [
          "    assert.notEqual(counterEl, null);",
          "    assertNotNullOrUndefined(counterEl);",
          "",
          "    const text = await (await counterEl.getProperty('innerText')).jsonValue();",
          "",
          "    assert.equal(text, '4');",
          "",
        ],
        "column": 25,
        "fromRowNumber": 108,
        "row": 111,
        "sourceFileRelativePath": "lib/sanity-integration-playwright.test.ts",
        "toRowNumber": 114,
      },
      "text": "innerText",
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 20,
      "fnName": "click",
      "index": 8,
      "mhtmlFile": "8.document.mhtml",
      "name": "click \\"#reset\\"",
      "pageId": 0,
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "8.page-click.jpg",
      "selector": "#reset",
      "startTimestamp": 19,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_Page_click",
            "line": 196,
          },
          Object {
            "column": 5,
            "file": "lib/sanity-integration-playwright.test.ts",
            "function": "Object.<anonymous>",
            "line": 115,
          },
        ],
        "codeLines": Array [
          "",
          "    assert.equal(text, '4');",
          "",
          "    await playedPage.click('#reset');",
          "",
          "    const text2 = await (await counterEl.getProperty('innerText')).jsonValue();",
          "",
        ],
        "column": 5,
        "fromRowNumber": 112,
        "row": 115,
        "sourceFileRelativePath": "lib/sanity-integration-playwright.test.ts",
        "toRowNumber": 118,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 22,
      "fnName": "getProperty",
      "index": 9,
      "mhtmlFile": "9.document.mhtml",
      "name": "getProperty \\"innerText\\" on \\"#counter\\"",
      "pageId": 0,
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "9.elementhandle-getProperty.jpg",
      "selector": "#counter",
      "startTimestamp": 21,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_ElementHandle_getProperty",
            "line": 196,
          },
          Object {
            "column": 26,
            "file": "lib/sanity-integration-playwright.test.ts",
            "function": "Object.<anonymous>",
            "line": 117,
          },
        ],
        "codeLines": Array [
          "",
          "    await playedPage.click('#reset');",
          "",
          "    const text2 = await (await counterEl.getProperty('innerText')).jsonValue();",
          "",
          "    assert.equal(text2, '0');",
          "",
        ],
        "column": 26,
        "fromRowNumber": 114,
        "row": 117,
        "sourceFileRelativePath": "lib/sanity-integration-playwright.test.ts",
        "toRowNumber": 120,
      },
      "text": "innerText",
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 24,
      "fnName": "$$",
      "index": 10,
      "mhtmlFile": "10.document.mhtml",
      "name": "$$ \\"button\\"",
      "pageId": 0,
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "10.page-$$.jpg",
      "selector": "button",
      "startTimestamp": 23,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 30,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "Proxy.returningElementHandlesViaPromiseWrappedFunction_Page_$$",
            "line": 159,
          },
          Object {
            "column": 24,
            "file": "lib/sanity-integration-playwright.test.ts",
            "function": "Object.<anonymous>",
            "line": 121,
          },
        ],
        "codeLines": Array [
          "",
          "    assert.equal(text2, '0');",
          "",
          "    const allButtons = await playedPage.$$('button');",
          "",
          "    await allButtons[0].click();",
          "",
        ],
        "column": 24,
        "fromRowNumber": 118,
        "row": 121,
        "sourceFileRelativePath": "lib/sanity-integration-playwright.test.ts",
        "toRowNumber": 124,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 26,
      "fnName": "click",
      "index": 11,
      "mhtmlFile": "11.document.mhtml",
      "name": "click \\"button[0]\\"",
      "pageId": 0,
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "11.elementhandle-click.jpg",
      "selector": "button[0]",
      "startTimestamp": 25,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_ElementHandle_click",
            "line": 196,
          },
          Object {
            "column": 5,
            "file": "lib/sanity-integration-playwright.test.ts",
            "function": "Object.<anonymous>",
            "line": 123,
          },
        ],
        "codeLines": Array [
          "",
          "    const allButtons = await playedPage.$$('button');",
          "",
          "    await allButtons[0].click();",
          "",
          "    const text3 = await (await counterEl.getProperty('innerText')).jsonValue();",
          "",
        ],
        "column": 5,
        "fromRowNumber": 120,
        "row": 123,
        "sourceFileRelativePath": "lib/sanity-integration-playwright.test.ts",
        "toRowNumber": 126,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 28,
      "fnName": "getProperty",
      "index": 12,
      "mhtmlFile": "12.document.mhtml",
      "name": "getProperty \\"innerText\\" on \\"#counter\\"",
      "pageId": 0,
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "12.elementhandle-getProperty.jpg",
      "selector": "#counter",
      "startTimestamp": 27,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_ElementHandle_getProperty",
            "line": 196,
          },
          Object {
            "column": 26,
            "file": "lib/sanity-integration-playwright.test.ts",
            "function": "Object.<anonymous>",
            "line": 125,
          },
        ],
        "codeLines": Array [
          "",
          "    await allButtons[0].click();",
          "",
          "    const text3 = await (await counterEl.getProperty('innerText')).jsonValue();",
          "",
          "    assert.equal(text3, '1');",
          "",
        ],
        "column": 26,
        "fromRowNumber": 122,
        "row": 125,
        "sourceFileRelativePath": "lib/sanity-integration-playwright.test.ts",
        "toRowNumber": 128,
      },
      "text": "innerText",
      "unhandledExceptions": Array [],
    },
  ],
}
`;
